Bug Fixes and Undo

Hakazaba opened the pull request:
Duplication can now be undone, and a couple of minor bugs have been fixed to do with zsort and node names.

Hakazaba committed:
Bug Fixes and Undo

Duplication can now be undone, and a couple of minor bugs have been fixed to do with zsort and node names.

r888800009 commented on source/creator/panels/nodes.d:
```diff
@@ -175,13 +175,7 @@ protected:
-                            l=l[0..$-1];
-                        }
+                        incAddChildWithHistory(x,n.parent,n.name~(" Copy"));
                         incActivePuppet().rescanNodes();
```
I guess the `IncActivePuppet().rescanNodes();` can be removed?
https://github.com/Inochi2D/inochi-creator/blob/85c2f6beb62f023b1427875155f1951e157a5fb8/source/creator/actions/node.d#L319-L333

Hakazaba replied:
Removed, along with a little transform bug this introduced

r888800009 commented:
Some other suggestions, maybe you can confirm whether the `Node` class and its inheritance has functions such as `serialization` or `deepcopy`, the data structure should be maintained by class, you can refer to
https://github.com/Inochi2D/inochi2d/tree/v0_8/source/inochi2d/core/nodes

Hakazaba committed:
Transform bug

Inc add child was clearing the transform of the node

Hakazaba commented:
> Some other suggestions, maybe you can confirm whether the `Node` class and its inheritance has functions such as `serialization` or `deepcopy`, the data structure should be maintained by class, you can refer to https://github.com/Inochi2D/inochi2d/tree/v0_8/source/inochi2d/core/nodes

There are some cases where i don't think we want a deep copy. For example, references to masks(Which still has some undesirable behavior for internal masks to the duplicated tree at the moment) and especially texture references, which is the main goal of this. So that we don't need duplicates in the texture atlass.

That being said, I'm not sure how to interact with serialization.

Hakazaba committed:
Mesh group duplication

Meshes weren't being duplicated correctly

Hakazaba committed:
Added compatible pairing copy

Would prefer to remove the origonal paste as i don't really see its usecase, so an ugly workaround to use axis 2 was added for a different function.

This version is nessisary to get a clean mirror on a negative scaled object, or any copied object. The origional paste or mirror will always cause imperfections.

Hakazaba committed:
Change of max zoom

This is possibly a matter of opinion, but I find it painful to sort out vert bundles at the previous zoom level and find this more comfortable
